[GSoC2024] added accumulation of confusion matrix across various jobs

Viditagarwal7479 opened the pull request:
Added accumulation of confusion matrix across various jobs under the same task.

### Motivation and context
Fixes https://github.com/opencv/cvat/issues/7424

### How has this been tested?
On the backup task containing 10 jobs [task_task.with.10.jobs_backup_2024_03_11_15_48_27.zip](https://github.com/opencv/cvat/files/14561155/task_task.with.10.jobs_backup_2024_03_11_15_48_27.zip) as mentioned in this [issue comment](https://github.com/opencv/cvat/issues/7424#issuecomment-1988787512)

### Checklist
- [x] I submit my changes into the `develop` branch
~- [ ] I have created a changelog fragment <!-- see top comment in CHANGELOG.md -->~
~- [ ] I have updated the documentation accordingly~
~- [ ] I have added tests to cover my changes~
- [x] I have linked related issues (see [GitHub docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
~- [ ] I have increased versions of npm packages if it is necessary
  ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
  [cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning),
  [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and
  [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))~

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.

Viditagarwal7479 commented:
I am attaching the quality report on the above backup task, for better understanding:
Quality Report before this PR: [quality-report-task_52-165.json](https://github.com/opencv/cvat/files/14606017/quality-report-task_52-165.json)
Quality Report after this PR: [quality-report-task_52-176.json](https://github.com/opencv/cvat/files/14606020/quality-report-task_52-176.json)

zhiltsov-max commented:
Hi, thank you for fixing the problem! Could you please also add a test for this to protect the code from regressions? It should be in [this file](https://github.com/opencv/cvat/blob/develop/tests/python/rest_api/test_quality_control.py).

Viditagarwal7479 commented:
As mentioned in the [documentation](https://github.com/opencv/cvat/tree/develop/tests/python#how-to-upgrade-testing-assets) on creating custom test cases.
> Add new objects (e.g. issues, comments, tasks, projects)

How to add new objects?
After starting the testing service,
1. How to login as the `admin1` user? I cannot find the needed password.
2. On reading the documentation I thought that I would open the CVAT server locally from the above (`pytest tests/python/ --start-services`) started service. It would fetch the tasks and jobs from the testing database and then I have to add a task as per my use case and then dump the contents. But On registering as a new user, I cannot see any existing tasks and jobs from the testing db.

Viditagarwal7479 commented:
In the current, [quality_report.json](https://github.com/opencv/cvat/blob/develop/tests/python/shared/assets/quality_reports.json) dump, confusion_matrix isn't dumped, and while going through the subsequent function calls, I wasn't able to get through any part where it could be defined to what parameters to include in the dump.

zhiltsov-max commented:
> How to add new objects?

Typically, we start the CVAT services with `pytest --start-services --rebuild ./tests/python/`, then we do necessary updates on the started instance in the browser UI (al localhost:8080, all the users there have the [same password](https://github.com/opencv/cvat/blob/f513aa162ba2abfeb6625408ae4b35ae6223c180/tests/python/shared/utils/config.py#L13)) or directly on the server. Then we dump the updated DB with one or both of [these commands](https://github.com/opencv/cvat/blob/develop/tests/python/README.md#how-to-backup-db-and-data-volume) and update the testing assets with [`(cd tests/ && python shared/utils/dump_objects.py)`](https://github.com/opencv/cvat/blob/develop/tests/python/README.md#how-to-update-json-files-in-the-assets-directory) if necessary.

Keep in mind that adding new assets is optional. You can always reuse any of the existing ones, if it fits the test, or construct it dynamically in the test, if needed. In most cases there is no need to register new users, and I believe, this PR also doesn't require this.

> In the current, [quality_report.json](https://github.com/opencv/cvat/blob/develop/tests/python/shared/assets/quality_reports.json) dump, confusion_matrix isn't dumped, and while going through the subsequent function calls, I wasn't able to get through any part where it could be defined to what parameters to include in the dump.

Yes, we don't send full reports in the endpoint these assets are used with. Please find an example of getting a full report [here](https://github.com/opencv/cvat/blob/develop/tests/python/rest_api/test_quality_control.py#L310). I suggest that new tests are added into [this class](https://github.com/opencv/cvat/blob/f513aa162ba2abfeb6625408ae4b35ae6223c180/tests/python/rest_api/test_quality_control.py#L1079).

Viditagarwal7479 commented:
Thank you @zhiltsov-max.
When I am trying to use the above mentioned password it's showing invalid credentials
In the GUI:
<img width="1224" alt="image" src="https://github.com/opencv/cvat/assets/72168180/0aa9ec9a-3231-4662-8326-b7ea5df05291">
When I am trying to do the same through API request:
<img width="253" alt="image" src="https://github.com/opencv/cvat/assets/72168180/346eeda2-24bb-4c2c-aad9-c5069f9fe1a3">

Viditagarwal7479 commented:
Should I create a separate dump file which includes the complete `quality_report` (including the confusion matrix) or should I update the dump of the current `quality_report.json` itself?

zhiltsov-max commented:
> When I am trying to use the above mentioned password it's showing invalid credentials. In the GUI:

It's possible after the tests have runned. Please try to relaunch with `pytest --stop-services ./tests/python && pytest --start-services ./tests/python`.

> Should I create a separate dump file which includes the complete quality_report (including the confusion matrix) or should I update the dump of the current quality_report.json itself?

I don't think it makes sense now. Probably, you can just create a test that computes and downloads a report and checks its values.

Viditagarwal7479 commented:
Thank you, that's so silly of me. It began working now.

> I don't think it makes sense now. Probably, you can just create a test that computes and downloads a report, and checks its values.

To check whether the computed values match the expected ones, we need to store the expected values somewhere. So, where can we store the expected confusion matrix corresponding to this new job?

zhiltsov-max commented:
> To check whether the computed values match the expected ones, we need to store the expected values somewhere. So, where can we store the expected confusion matrix corresponding to this new job?

I agree, but we only need to check the matrices, and you probably need to construct a task with several jobs to reproduce the problem. I think we can do all of this in the test.

Viditagarwal7479 commented:
Oh, ok, that makes sense; I will add a sample test case and compare the computed test case with the hardcoded value of the confusion matrix.

Viditagarwal7479 commented:
1. https://github.com/opencv/cvat/blob/4851640f4ce5b4193e0d460611d3f35fe8a1ee0b/tests/python/rest_api/test_jobs.py#L83 In such test cases there should also be a condition to check whether the selected task has a ground truth job or not. (`and not any(j for j in jobs if j["task_id"] == t["id"] and j["type"] == "ground_truth")`)
2. In the test [`test_can_get_gt_job_frame`](https://github.com/opencv/cvat/blob/4851640f4ce5b4193e0d460611d3f35fe8a1ee0b/tests/python/rest_api/test_jobs.py#L749) at [this line](https://github.com/opencv/cvat/blob/4851640f4ce5b4193e0d460611d3f35fe8a1ee0b/tests/python/rest_api/test_jobs.py#L768) when it's trying to get or create a gt_job which only includes frames having `job_frame_ids`. In that function, it ensures that the frames with `job_frame_ids` are only used while creating a new gt_job, but the same doesn't happen when the selected task has a gt_job. Thus, in such a case, we should add a check that only picks up tasks that don't have a gt_job corresponding to them. In the long run, we should correct the method which fetches or creates gt_job here.
3. Same as mentioned above happens for the test [`test_can_get_gt_job_meta`](https://github.com/opencv/cvat/blob/4851640f4ce5b4193e0d460611d3f35fe8a1ee0b/tests/python/rest_api/test_jobs.py#L580)
4. I have updated the database and the dump of JSON files, but in this, the [`quality_reports.json`](https://github.com/opencv/cvat/blob/4851640f4ce5b4193e0d460611d3f35fe8a1ee0b/tests/python/shared/assets/quality_reports.json) wasn't updated due to which the new task isn't included in that json which causes failure due to empty list in this test [`test_modified_task_produces_different_metrics`](https://github.com/opencv/cvat/blob/4851640f4ce5b4193e0d460611d3f35fe8a1ee0b/tests/python/rest_api/test_quality_control.py#L1129)
5. I am not able to figure out what does this test intend to do so I am not able to figure out why it's failing. Can you please help in this. [`test_two_project_webhooks_intersection`](https://github.com/opencv/cvat/blob/4851640f4ce5b4193e0d460611d3f35fe8a1ee0b/tests/python/rest_api/test_webhooks_sender.py#L205)

Viditagarwal7479 commented:
If you also consider that the above analysis is correct for points 1, 2 and 3 then should I create a new issue for them and create a PR fixing them?

zhiltsov-max commented:
@Viditagarwal7479,

> If you also consider that the above analysis is correct for points 1, 2 and 3 then should I create a new issue for them and create a PR fixing them?

Yes, please send a separate PR, if you see how the existing tests can be improved.

> 4. I have updated the database and the dump of JSON files, but in this, the [quality_reports.json](https://github.com/opencv/cvat/blob/acb68a54d75456338802e2774ef0ffd388bad761/tests/python/shared/assets/quality_reports.json) wasn't updated

After the DB is updated and the server is restarted, it should have been updated automatically after calling `dump_objects.py`. Did you execute this script?

Also, as a side note, typically we clean out unnecessary `data.json` updates, such as new auth tokens and records, if that's not the point of the update.

> 5. I am not able to figure out what does this test intend to do so I am not able to figure out why it's failing. Can you please help in this. [test_two_project_webhooks_intersection](https://github.com/opencv/cvat/blob/acb68a54d75456338802e2774ef0ffd388bad761/tests/python/rest_api/test_webhooks_sender.py#L205)

I think it checks that if there are 2 webhooks registered on a projects, and these webhooks both listen to the task creation event, they both should be triggered once a task is created in the project. I don't think this test is related to the changes, so it should not be affected. One possible reason is that some tests can fail if you execute tests successively without restarting the test server. Please try to stop the test containers before running the tests again.

Viditagarwal7479 commented:
> After the DB is updated and the server is restarted, it should have been updated automatically after calling dump_objects.py. Did you execute this script?

Yes, that's why you can see in the commit [`added tests and updated test assets`](https://github.com/opencv/cvat/pull/7604/commits/b234b1f114f2523be48c8ae4d275c781d80da81f) many of the JSON files have been updated.

> Also, as a side note, typically we clean out unnecessary data.json updates, such as new auth tokens and records, if that's not the point of the update.

Ohh, ok, thank you. In such a case, I shall revert the change in `data.json`.

> One possible reason is that some tests can fail if you execute tests successively without restarting the test server. Please try to stop the test containers before running the tests again.

I tried this, but it still keeps failing (locally), whereas this test case isn't executed in the above `Helm tests`

Viditagarwal7479 commented:
Also, in the Helm tests, which GitHub actions are running. There,
1. [`tests/python/rest_api/test_jobs.py::TestListJobs::test_admin_list_jobs`](https://github.com/opencv/cvat/blob/f513aa162ba2abfeb6625408ae4b35ae6223c180/tests/python/rest_api/test_jobs.py#L808)
2. [`tests/python/rest_api/test_jobs.py::TestJobsListFilters::test_can_use_simple_filter_for_object_list`](https://github.com/opencv/cvat/blob/f513aa162ba2abfeb6625408ae4b35ae6223c180/tests/python/rest_api/test_jobs.py#L857)
3. [`tests/python/rest_api/test_tasks.py::TestGetTasks::test_can_remove_owner_and_fetch_with_sdk`](https://github.com/opencv/cvat/blob/f513aa162ba2abfeb6625408ae4b35ae6223c180/tests/python/rest_api/test_tasks.py#L203)

And a few more tests are failing, which are passing on my local machine. In such cases, I suspect that while running the above tests, GitHub actions might not be using the updated json files.

Viditagarwal7479 commented:
Hey @zhiltsov-max ,
> Yes, please send a separate PR, if you see how the existing tests can be improved. I think, the changes you're suggesting make sense.

I have created a [new issue](https://github.com/opencv/cvat/issues/7622) regarding the same and a [PR](https://github.com/opencv/cvat/pull/7623) to fix too.

Viditagarwal7479 commented:
Regarding the `quality_report.json` not being updated after updating the tests and dumping,

Due to unknown reasons (I cannot figure out why), the test server `localhost:8080` UI application isn't computing the quality reports. I observed this on the new task (which I added in this PR), on a simple new dummy task, and by updating the earlier only task with a gt_job. But this works fine when querying through API or testing this on the normal build `localhost:3000`.
Would you please suggest something based on this observation?

<img width="874" alt="image" src="https://github.com/opencv/cvat/assets/72168180/6562b2ba-d508-40ac-acf8-6164c092a857">

zhiltsov-max commented:
Sorry, quality reports are computed once in some time after the last task update by default as it's a costly operation for big tasks. As an admin, you can trigger computation manually using the POST `/api/quality/reports/` endpoint in `http://localhost:8080/api/swagger/`. You'll need to specify the task id and, optionally, login. Another option is to change the default computation period by changing this [variable](https://github.com/opencv/cvat/blob/acb68a54d75456338802e2774ef0ffd388bad761/cvat/apps/quality_control/default_settings.py#L7) either as a container environment variable or setting in `base.py`.

Viditagarwal7479 commented:
Thank you @zhiltsov-max, with your help I computed the quality report and have added the new commit. Once the PR related to tests gets merged hopefully this PR also would pass all the test cases and would be good to merge.

This information about making the manual POST request is missing in the docs, should I add this information [here](https://github.com/opencv/cvat/blob/f513aa162ba2abfeb6625408ae4b35ae6223c180/tests/python/README.md)?

zhiltsov-max commented on cvat/apps/quality_control/quality_reports.py:
```diff
@@ -2394,6 +2401,15 @@ def _compute_task_report(
                 task_frame_results[frame_id] = task_frame_result
 
+        if confusion_matrices:
+            confusion_matrix = np.sum(confusion_matrices, axis=0)
```
As we discovered in the parallel [PR](https://github.com/opencv/cvat/pull/7619), this method is less efficient now that simple accumulation during iteration. Could you please update the code?

zhiltsov-max commented on tests/python/shared/assets/cvat_db/data.json:
```diff
@@ -36,7 +36,7 @@
   "fields": {
     "password": "md5$Ba6a4O5ImOYiZ04wBFMtMQ$d1b00970c835fcf12c17240f05f8257e",
-    "last_login": "2023-03-27T19:06:41.550Z",
+    "last_login": "2024-03-16T18:07:00.957Z",
```
Please clean up unnecessary changes.

zhiltsov-max commented on tests/python/shared/assets/cvat_db/data.json:
```diff
@@ -565,6 +565,14 @@
     "expire_date": "2022-03-10T20:44:57.190Z"
   }
 },
+{
+  "model": "sessions.session",
+  "pk": "hjh4pefasebs5qnwh27xm50uvzzv1txt",
+  "fields": {
+    "session_data": ".eJxVjMsOwiAURP-FtSFALy-X7v0GAtyLVA1NSrsy_ruSdKG7yZwz82Ih7lsNe6c1zMjOTLLTb5diflAbAO-x3Rael7atc-JD4Qft_LogPS-H-3dQY69jbWUBIaUHI7TSpIw3aqLkELTJmL1FVdBZrxOWb9YgivTKOEeTNQDs_QG87TbV:1rlYQv:8dWFCbJmXJ3C8TSo3Ic_C4Xfgrgsa6nmPzOEZFv39wY",
+    "expire_date": "2024-03-30T18:07:01.098Z"
+  }
+},
```
Please clean up unnecessary changes.

zhiltsov-max commented on tests/python/shared/assets/cvat_db/data.json:
```diff
@@ -791,11 +799,11 @@
   "model": "organizations.organization",
   "pk": 1,
   "fields": {
+    "created_date": "2021-12-14T18:45:40.172Z",
+    "updated_date": "2021-12-14T18:45:40.172Z",
```
Please clean up unnecessary changes.

zhiltsov-max commented:
> This information about making the manual POST request is missing in the docs, should I add this information [here](https://github.com/opencv/cvat/blob/f513aa162ba2abfeb6625408ae4b35ae6223c180/tests/python/README.md)?

It doesn't seem to be related to tests. It could be added in administration docs or in the quality estimation manual, but I'm not sure it will be helpful there, as regular users won't have it available. API docs have this endpoint mentioned.

Viditagarwal7479 commented:
Hello @zhiltsov-max, there are multiple more lines in `data.json` which are shown as changes, but they are just reformatting in which the order of keys is changed. I reverted the whole `data.json` file, which caused some tests to fail, so I have only fixed the regions you mentioned.

Viditagarwal7479 commented:
> > This information about making the manual POST request is missing in the docs, should I add this information [here](https://github.com/opencv/cvat/blob/f513aa162ba2abfeb6625408ae4b35ae6223c180/tests/python/README.md)?
> 
> It doesn't seem to be related to tests. It could be added in administration docs or in the quality estimation manual, but I'm not sure it will be helpful there, as regular users won't have it available. API docs have this endpoint mentioned.

Whenever any developer adds a new test case that requires computing the quality report, then until they don't trigger this API call, the updated database and JSON files will be incomplete.

Shouldn't how to add new tests be present in the [docs about contributing](https://opencv.github.io/cvat/docs/contributing/)?

zhiltsov-max commented:
> Whenever any developer adds a new test case that requires computing the quality report, then until they don't trigger this API call, the updated database and JSON files will be incomplete.

> Shouldn't how to add new tests be present in the [docs about contributing](https://opencv.github.io/cvat/docs/contributing/)?

Probably, it can be added into the test readme in the asset updating section - https://github.com/opencv/cvat/blob/develop/tests/python/README.md#how-to-upgrade-testing-assets . It should also mention the same operation for analytics, as it uses the same system.

Viditagarwal7479 commented:
[`tests/python/rest_api/test_quality_control.py::TestQualityReportMetrics::test_settings_affect_metrics`](https://github.com/opencv/cvat/blob/fc54c47ed6be1ee856fe14549d0a89b5535dc0c5/tests/python/rest_api/test_qual(ity_control.py#L1099) is designed such that only to pass when the selected `task_id` is 22, as seen from this [line](https://github.com/opencv/cvat/blob/fc54c47ed6be1ee856fe14549d0a89b5535dc0c5/tests/python/rest_api/test_quality_control.py#L1100) due to which it would pass before updating the testing db. I also tried by using other `task_id's`, but it failed and passed on `task_id` 22 as that task might be designed such that the change in parameters would affect the quality_report.

zhiltsov-max commented:
> is designed such that only to pass when the selected task_id is 22, as seen from this [line](https://github.com/opencv/cvat/blob/fc54c47ed6be1ee856fe14549d0a89b5535dc0c5/tests/python/rest_api/test_quality_control.py#L1100) due to which it would pass before updating the testing db. I also tried by using other task_id's, but it failed and passed on task_id 22 as that task might be designed such that the change in parameters would affect the quality_report.

Well, yes, all the tests referring to the task id 22 are designed to work with this task. That task contains most, if not all, examples of the comparable annotations and checkable annotation errors.

Viditagarwal7479 commented:
> all the tests referring to the task id 22 are designed to work with this task

Task id 23 only passes `test_settings_affect_metrics[low_overlap_threshold]`, not any other test in this class. Also, in this class, we already pass `task_id` as 22 in all the tests but only use that in a few. So, why not just update the test to use the parameter `task_id` passed into the tests?

zhiltsov-max commented:
I don't understand the suggestion. If the idea is to change the new test from using a class variable into a function parameter, I see no problems in doing this.

zhiltsov-max commented:
> Ah, I am trying to say that we should fix task_id as 22 in this test [tests/python/rest_api/test_quality_control.py::TestQualityReportMetrics::test_settings_affect_metrics](https://github.com/opencv/cvat/blob/fc54c47ed6be1ee856fe14549d0a89b5535dc0c5/tests/python/rest_api/test_quality_control.py#L1099) and other tests in this class, preferably as that tests are testing the platform not individual tests, so this should work fine.

> As you mentioned in the previous comment, task_id, 23 should also pass this test. I was mentioning that it isn't passing it.

Yes, if you're checking some general properties of objects - sure. But when you're testing something specific, you can't expect any chosen property to be observed on a random test sample. You need to build specific samples to be sure you can expect the functionality to work or bugs to be reproduced. Specifically, you can't expect a test called, say, `test_settings_affect_metrics`, to change the `min_overlap` setting on a task to some value, so that the task automatically  produces changed quality metrics, unless you specifically designed this task to do so. If you have an algorithm to build such test tasks or a set of pre-constructed test samples - yes, it can work with any of them.

Viditagarwal7479 commented:
Makes sense, so should I create a separate PR for this or make changes and commit to this PR itself?

zhiltsov-max commented on tests/python/rest_api/test_quality_control.py:
```diff
@@ -1202,3 +1203,27 @@ def test_settings_affect_metrics(
         new_report = self.create_quality_report(admin_user, task_id)
         assert new_report["summary"]["conflict_count"] != old_report["summary"]["conflict_count"]
+
+    @pytest.mark.usefixtures("restore_db_per_class")
```
```suggestion
```

This class already restores the DB before each test.

zhiltsov-max commented:
I think `test_settings_affect_metrics` should be updated to actually use the function parameter passed. It's okay to fix this in this PR.

Viditagarwal7479 commented:
https://github.com/opencv/cvat/pull/7640 in this PR today, the testing db and other JSON files have been updated, so there are merge conflicts here. I think so,
I would need to overwrite my changes with theirs, add the new test task again and then update the dump!

Viditagarwal7479 commented:
Hey @zhiltsov-max,
Now, only two tests are failing, and I cannot understand why they are failing.
1. I started the docker server
2. Added the tests and manually triggered the API to calculate the quality report
3. dumped the updated testing database and data.json
4. stopped the server as otherwise we would further get an authentication error
5. I started the server again as we dumped in the previous start thus, now we start just where we had left last time
6. Dumped and updated the JSON files

Still, some tests, unaffected by the changes I made, are failing. Also in the diff report, the date of change is of 2023, not even 2024

zhiltsov-max commented on tests/python/rest_api/test_quality_control.py:
```diff
@@ -1097,7 +1098,7 @@ def test_report_summary(self, task_id, tasks, jobs, quality_reports):
 
     def test_unmodified_task_produces_the_same_metrics(self, admin_user, quality_reports):
-        old_report = max((r for r in quality_reports if r["task_id"]), key=lambda r: r["id"])
+        old_report = next(r for r in quality_reports if r["task_id"] == self.demo_task_id)
```
Why was it changed to `next()`? It's only supposed to work with the latest task report, not any report.

Viditagarwal7479 replied:
In this test class, most tests were using the `task_id` 22, so thought this should also use it, so changed it accordingly. Shouldn't all tests pass this since the unmodified tasks should produce the same metrics?

Viditagarwal7479 replied:
After reverting to the original state to test it on the latest task, it's still failing.

zhiltsov-max replied:
I was only saying about the `max` function, not the `==` part. We need to use the latest task report, not any task report.

Viditagarwal7479 replied:
Ahh! Here the value being compared by max funding is report id not task Id. Due to this mistake I thought there will be a single report only so no need for max.

zhiltsov-max commented:
About the failing `TestSimpleQualityConflictsFilters` test - seems like the expected values for the `task_id` filter are not filled correctly now, since this new task has been added. Please update the code [here](https://github.com/opencv/cvat/blob/ef67a5033128045acb74a54031ab03d66b5c2427/tests/python/rest_api/test_quality_control.py#L688) to be like this:
```python
task_id = self._find_valid_field_value(self.report_samples, field_path=['task_id'])
task_reports = [r for r in self.report_samples if r["task_id"] == task_id]
```
This should fix the problem.

Viditagarwal7479 commented:
Done! Finally, I hope this becomes the last commit of this PR.

zhiltsov-max commented on tests/python/shared/assets/cvat_db/data.json:
```diff
@@ -394,7 +394,7 @@
   "fields": {
     "password": "md5$rKclttm53lRHC0rITF48uu$7ac59be88ed2a5de4b513edbef629258",
-    "last_login": "2022-03-05T09:37:01.636Z",
+    "last_login": "2024-03-21T20:49:58.088Z",
```
Please clean up the file.

```suggestion
```

zhiltsov-max commented on tests/python/shared/assets/users.json:
```diff
@@ -62,7 +62,7 @@
       "is_staff": true,
       "is_superuser": true,
-      "last_login": "2022-03-05T09:37:01.636000Z",
+      "last_login": "2024-03-21T20:49:58.088000Z",
```
```suggestion
    "last_login": "2022-03-05T09:37:01.636000Z",
```

zhiltsov-max replied:
Yes, I understand, but it's not really needed to be updated. Changing it only introduces extra merge conflicts later.

Viditagarwal7479 replied:
Hello, just a small follow-up on this, if these particular changes aren't needed. So, shouldn't we consider changing the dumping JSON code so as not to dump such values? Each time we look through a dump, it might be tiring to revert the unnecessary parts.

Viditagarwal7479 replied:
This goes for both `data.json` and `user.json`

zhiltsov-max replied:
I think, it can be added with as a default option with a parameter to dump everything.

Viditagarwal7479 replied:
Do you mean to say, like a flag `-i` to indicate only ones needed in the test to be updated, like what we wanted here? Or else dump everything.

Viditagarwal7479 replied:
Ah, that makes sense. Should I open an issue related to this?

zhiltsov-max replied:
It doesn't seem like a real issue to me, but you can create it, if you see it's a problem.

zhiltsov-max commented:
Thank you for contributing into the project!

zhiltsov-max commented:
Please send a PR with the changelog entry added.

Viditagarwal7479 commented:
> Please send a PR with the changelog entry added.

This is a bug fix, so what should I enter in the changelog?

zhiltsov-max commented:
Please find message examples in CHANGELOG.md or https://github.com/opencv/cvat/releases.

Viditagarwal7479 commented:
Thanks, I have created the changelog.md in this new PR.
